transports/dns/src/lib.rs

```diff
@@ -785,4 +795,126 @@ mod tests {
             rt.block_on(run(tokio::Transport::custom(CustomTransport, config, opts)));
         }
     }
+
+    #[test]
+    fn aggregated_dial_errors() {
+        let _ = tracing_subscriber::fmt()
+            .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
+            .try_init();
+
+        #[derive(Clone)]
+        struct AlwaysFailTransport;
+
+        impl libp2p_core::Transport for AlwaysFailTransport {
+            type Output = ();
+            type Error = std::io::Error;
+            type ListenerUpgrade = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+            type Dial = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+            fn listen_on(
+                &mut self,
+                _id: ListenerId,
+                _addr: Multiaddr,
+            ) -> Result<(), TransportError<Self::Error>> {
+                unimplemented!()
+            }
+
+            fn remove_listener(&mut self, _id: ListenerId) -> bool {
+                false
+            }
+
+            fn dial(
+                &mut self,
+                addr: Multiaddr,
+                _: DialOpts,
+            ) -> Result<Self::Dial, TransportError<Self::Error>> {
+                // Every dial attempt fails with an error that includes the address.
+                Ok(Box::pin(future::ready(Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    format!("dial error for {}", addr),
+                )))))
+            }
+
+            fn poll(
+                self: Pin<&mut Self>,
+                _cx: &mut Context<'_>,
+            ) -> Poll<TransportEvent<Self::ListenerUpgrade, Self::Error>> {
+                unimplemented!()
+            }
+        }
+
+        async fn run_test<T, R>(mut transport: super::Transport<T, R>)
+        where
+            T: Transport + Clone + Send + Unpin + 'static,
+            T::Error: Send,
+            T::Dial: Send,
+            R: Clone + Send + Sync + Resolver + 'static,
+        {
+            let dial_opts = DialOpts {
+                role: Endpoint::Dialer,
+                port_use: PortUse::Reuse,
+            };
+
+            // This address requires DNS resolution, yielding two IP addresses,
+            // forcing two dial attempts. Both fail.
+            let addr: Multiaddr = "/dns/youtube-ui.l.google.com/tcp/1234".parse().unwrap();
```

Comment:
```suggestion
            let addr: Multiaddr = "/dnsaddr/bootstrap.libp2p.io".parse().unwrap();
```
I think the bootstrap address also yields multiple records.